x-pack/plugins/security_solution/public/management/cypress/e2e/artifact_tabs_in_policy_details.cy.ts

```diff
@@ -0,0 +1,221 @@
+  cy.get(`#${tabId}`).click();
+};
+
+const visitPolicyDetailsPage = () => {
+  cy.visit('/app/security/administration/policy');
+  cy.getBySel('policyNameCellLink').eq(0).click({ force: true });
+  cy.getBySel('policyDetailsPage').should('exist');
+};
+
+describe('Artifact tabs in Policy Details page', () => {
+  before(() => {
+    login();
+    loadEndpointDataForEventFiltersIfNeeded();
+    removeAllArtifacts();
+  });
+
+  after(() => {
+    removeAllArtifacts();
+  });
+
+  for (const testData of getArtifactsListTestsData()) {
+    describe(`${testData.title} tab`, () => {
+      it(`[NONE] User cannot see the tab for ${testData.title}`, () => {
+        loginWithPrivilegeNone(testData.privilegePrefix);
+        visitPolicyDetailsPage();
+
+        cy.get(`#${testData.tabId}`).should('not.exist');
+      });
+
+      context(`Given there are no ${testData.title} entries`, () => {
+        it(`[READ] User CANNOT add ${testData.title} artifact`, () => {
+          loginWithPrivilegeRead(testData.privilegePrefix);
+          visitArtifactTab(testData.tabId);
+
+          cy.getBySel('policy-artifacts-empty-unexisting').should('exist');
+
+          cy.getBySel('unexisting-manage-artifacts-button').should('not.exist');
+        });
+
+        it(`[ALL] User can add ${testData.title} artifact`, () => {
+          loginWithPrivilegeAll();
+          visitArtifactTab(testData.tabId);
+
+          cy.getBySel('policy-artifacts-empty-unexisting').should('exist');
+
+          cy.getBySel('unexisting-manage-artifacts-button').should('exist').click();
+
+          const { formActions, checkResults } = testData.create;
+
+          performUserActions(formActions);
+
+          // Add a per policy artifact - but not assign it to any policy
+          cy.get('[data-test-subj$="-perPolicy"]').click(); // test-subjects are generated in different formats, but all ends with -perPolicy
+          cy.getBySel(`${testData.pagePrefix}-flyout-submitButton`).click();
+
+          // Check new artifact is in the list
+          for (const checkResult of checkResults) {
+            cy.getBySel(checkResult.selector).should('have.text', checkResult.value);
+          }
+
+          cy.getBySel('policyDetailsPage').should('not.exist');
+          cy.getBySel('backToOrigin').contains(/^Back to .+ policy$/);
+
+          cy.getBySel('backToOrigin').click();
+          cy.getBySel('policyDetailsPage').should('exist');
+        });
+      });
+
+      context(`Given there are no assigned ${testData.title} entries`, () => {
+        it(`[READ] User CANNOT Manage or Assign ${testData.title} artifacts`, () => {
+          loginWithPrivilegeRead(testData.privilegePrefix);
+          visitArtifactTab(testData.tabId);
+
+          cy.getBySel('policy-artifacts-empty-unassigned').should('exist');
+
+          cy.getBySel('unassigned-manage-artifacts-button').should('not.exist');
+          cy.getBySel('unassigned-assign-artifacts-button').should('not.exist');
```

Comment:
indeed, but I didn't see it necessary. instead, first I check if the container exists (`policy-artifacts-empty-unexisting` and `policy-artifacts-empty-unassigned`), then I check the corresponding buttons (`unexisting-*` and `unassigned-*`). what do you think?

Reply:
I see what you mean. Although, I expect a test to have all the assertions based on the given conditions encapsulated in one place. 

So, if I need to understand how a particular feature works I just need to read that one test block and glean everything I need. 

However, spreading the behavior in separate test blocks would mean I need to read all of the test blocks to fully understand what a feature is supposed to do.

Reply:
I think I see your point - but I am not sure where the line is about what to include amongst the `not.exist` assertions, and also they are easily be false positives, so I prefer to have as few as possible. Copying these lines to above would lead to copy them into every test case, and asserting for every possible not existing button in every case - I'd avoid this for sure. I believe this would make understanding the features even harder, not easier. What do you think of this?